### DISCUSSION TITLE
Is it a correct approach to use default library context for loading providers?

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#24177] [Q&A] [answered]
Hi All,

I wonder if it is correct to use `NULL` as library context to manually load configuration and providers in the default library context or not. For example, is the following code valid?
```
OSSL_PROVIDER *init(const char *providername)
{
        OSSL_PROVIDER *provider = NULL;
	if (providername) {
		provider = OSSL_PROVIDER_load(NULL, providername);
		if (!provider)
			NULL;
	}
	return provider;
}
```
I assume this code first loads default `openssl.cnf`, disables fallback, then loads additional provider into it if any name is passed.
I'm asking this because all examples in openssl will try load providers in an non-default context. I wonder if there any problem with loading it into default one.
And should I manually free all providers that I have loaded into default library context?

**Comment by levitte** [comment, marked as answer]
It's currently legitimate to use the default library context the way you do.  However, we *recommend* that you create your own library context.

**Comment by mattcaswell** [comment]
> And should I manually free all providers that I have loaded into default library context?

Manually free all those that you have manually loaded via `OSSL_PROVIDER_load` (you don't need to do this for any that have been loaded by config)
